### PR TITLE
[maistra-725] Add proxy logs

### PR DESCRIPTION
**Pull request by brian-avery**
This PR adds logs for must-gather based on istioctl proxy-status.

In the directory containing the control plane namespace, a new directory is created called controlplane. This contains a file called synchronization. The contents are shown below. Note that in 1.1, the status of each synchronization (cluster, listener, route, endpoint) can be found by comparing resource_sent and resource_acked. 
If:
* the two are equal: then Pilot and Envoy are synchronized.
* resource_sent is empty: config has not been sent
* otherwise this returns stale. 1.2+ have additional functionality when this changes to using time instead. 

```
[
    {
        "proxy": "reviews-v1-989d5ffdf-nvz5l.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "b73de5ee-8c72-449b-a23f-9d62a2fe4560",
        "cluster_acked": "b73de5ee-8c72-449b-a23f-9d62a2fe4560",
        "listener_sent": "cdd69f35-ede6-4a19-a94d-11c9f6db1ab0",
        "listener_acked": "cdd69f35-ede6-4a19-a94d-11c9f6db1ab0",
        "route_sent": "c0d622e2-ffa7-40ee-baa4-301a341c3232",
        "route_acked": "c0d622e2-ffa7-40ee-baa4-301a341c3232",
        "endpoint_sent": "4cdd75ba-3d7d-411e-ba99-a39f8a885e80",
        "endpoint_acked": "4cdd75ba-3d7d-411e-ba99-a39f8a885e80",
        "endpoint_percent": 50
    },
    {
        "proxy": "reviews-v3-757c4f7849-fk5f8.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "46db4a02-e925-4dbf-99c2-b1256199fda5",
        "cluster_acked": "46db4a02-e925-4dbf-99c2-b1256199fda5",
        "listener_sent": "8003932f-3290-4744-9cae-95b38779a467",
        "listener_acked": "8003932f-3290-4744-9cae-95b38779a467",
        "route_sent": "0da6d7cc-fcc8-4bb5-9b6f-b18ea54394b6",
        "route_acked": "0da6d7cc-fcc8-4bb5-9b6f-b18ea54394b6",
        "endpoint_sent": "c6ceef3b-f66b-4fbf-a650-ec806421b05c",
        "endpoint_acked": "c6ceef3b-f66b-4fbf-a650-ec806421b05c",
        "endpoint_percent": 50
    },
    {
        "proxy": "ratings-v1-d4db56d5b-c5cpd.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "057b72d4-d044-40b5-ab56-9c5f1ff2266d",
        "cluster_acked": "057b72d4-d044-40b5-ab56-9c5f1ff2266d",
        "listener_sent": "0f0c6b3a-e96d-4277-8b3a-509edf23bbe9",
        "listener_acked": "0f0c6b3a-e96d-4277-8b3a-509edf23bbe9",
        "route_sent": "3dc34e66-6168-46d1-8dd4-f445bbb1bf8f",
        "route_acked": "3dc34e66-6168-46d1-8dd4-f445bbb1bf8f",
        "endpoint_sent": "c3cdbc19-3a65-45ab-900c-2f187f6dfb38",
        "endpoint_acked": "c3cdbc19-3a65-45ab-900c-2f187f6dfb38",
        "endpoint_percent": 50
    },
    {
        "proxy": "reviews-v2-6ff8648d69-pbc5v.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "f34323ac-a727-4a51-b137-81c191673309",
        "cluster_acked": "f34323ac-a727-4a51-b137-81c191673309",
        "listener_sent": "c5d69454-0242-4e08-8ab1-d7ebed5a330e",
        "listener_acked": "c5d69454-0242-4e08-8ab1-d7ebed5a330e",
        "route_sent": "e0a79bbd-8310-4b7e-80ff-a624f669e6a7",
        "route_acked": "e0a79bbd-8310-4b7e-80ff-a624f669e6a7",
        "endpoint_sent": "9254db2e-9691-4f05-9806-de04737533b1",
        "endpoint_acked": "9254db2e-9691-4f05-9806-de04737533b1",
        "endpoint_percent": 50
    },
    {
        "proxy": "productpage-v1-7ccb674454-jx6vt.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "aeea9d47-461f-42e0-aa85-a7ec0ffcd080",
        "cluster_acked": "aeea9d47-461f-42e0-aa85-a7ec0ffcd080",
        "listener_sent": "6722b8fc-7fa5-4bfa-8ec8-e486ac0971dc",
        "listener_acked": "6722b8fc-7fa5-4bfa-8ec8-e486ac0971dc",
        "route_sent": "aeb64966-647a-478d-b2e6-b836de94c320",
        "route_acked": "aeb64966-647a-478d-b2e6-b836de94c320",
        "endpoint_sent": "027dff1d-23ca-4d42-b9cf-1d24f412dff6",
        "endpoint_acked": "027dff1d-23ca-4d42-b9cf-1d24f412dff6",
        "endpoint_percent": 50
    },
    {
        "proxy": "istio-egressgateway-5bf868b7c-bl8qm.istio-system",
        "proxy_version": "1.1.11",
        "cluster_sent": "f3f0579f-2c84-41fd-8577-c1bfad075745",
        "cluster_acked": "f3f0579f-2c84-41fd-8577-c1bfad075745",
        "listener_sent": "681f591a-f723-4968-bb2f-2fb8561a1a68",
        "listener_acked": "681f591a-f723-4968-bb2f-2fb8561a1a68",
        "endpoint_sent": "59285fe9-30be-4fca-b9b1-b626e6d636ef",
        "endpoint_acked": "59285fe9-30be-4fca-b9b1-b626e6d636ef",
        "endpoint_percent": 100
    },
    {
        "proxy": "istio-ingressgateway-56d5f5cddd-6nzwk.istio-system",
        "proxy_version": "1.1.11",
        "cluster_sent": "974d8644-542c-477f-8848-144b8cbc64b9",
        "cluster_acked": "974d8644-542c-477f-8848-144b8cbc64b9",
        "listener_sent": "3b5be3c1-9fc1-4d7f-bf4a-84f9a6ba62b3",
        "listener_acked": "3b5be3c1-9fc1-4d7f-bf4a-84f9a6ba62b3",
        "route_sent": "d3e6b4c9-1e1f-4b6b-b2e9-a779dacb43a8",
        "route_acked": "d3e6b4c9-1e1f-4b6b-b2e9-a779dacb43a8",
        "endpoint_sent": "a82f6ee3-e07f-4987-a5e2-55dd576c4ce5",
        "endpoint_acked": "a82f6ee3-e07f-4987-a5e2-55dd576c4ce5",
        "endpoint_percent": 100
    },
    {
        "proxy": "details-v1-5c88c7575b-rb5ls.bookinfo",
        "proxy_version": "1.1.11",
        "cluster_sent": "e546b98b-9828-4bce-ba63-4f89b208b5e5",
        "cluster_acked": "e546b98b-9828-4bce-ba63-4f89b208b5e5",
        "listener_sent": "b63c5f50-1577-47b9-9449-c72baeb82648",
        "listener_acked": "b63c5f50-1577-47b9-9449-c72baeb82648",
        "route_sent": "ecdb38f2-e5ce-475a-bb43-731a04a36dd7",
        "route_acked": "ecdb38f2-e5ce-475a-bb43-731a04a36dd7",
        "endpoint_sent": "36be2f80-8196-4f21-8792-e5f37b7b13ea",
        "endpoint_acked": "36be2f80-8196-4f21-8792-e5f37b7b13ea",
        "endpoint_percent": 50
    }
]

```
Each pod in the control plane (both istio-system and member pods) that has an attached proxy also generates a file called pilotConfiguration containing the xDS configuration as seen by Pilot, and one called envoyConfiguration containing the xDS configuration as seen by Envoy. Upstream compares the clusters, listeners, and routes in order to determine if everything is synchronized and shows the expected for anything that doesn't match. We just dump it. 

An example of the output from these can be seen below: 
[envoyConfiguration.txt](https://github.com/Maistra/istio-must-gather/files/3534833/envoyConfiguration.txt)
[pilotConfiguration.txt](https://github.com/Maistra/istio-must-gather/files/3534834/pilotConfiguration.txt)

### CODE REVIEW

File: gather_istio [hunk 1]
```diff
@@ -77,40 +78,89 @@ function getCRDs() {
...
+        fi
+    done
+}
+
+function main() {
     echo
-    echo "Dumping CRD ${crd}..."
-    ${MUST_GATHER} inspect --all-namespaces ${crd}
-  done
+    echo "Brian Executing Istio gather script"
```
Comment (dgn) [inline gather_istio]: ?

File: gather_istio [hunk 2]
```diff
@@ -77,40 +78,89 @@ function getCRDs() {
...
+    echo
+
+		echo "Collecting Envoy Configuration"
```
Comment (dgn) [inline gather_istio]: there's a weird mix of tabs and spaces in this file

File: gather_istio [hunk 3]
```diff
@@ -15,6 +15,7 @@
 # limitations under the License.
 
 MUST_GATHER=${MUST_GATHER:-/usr/bin/openshift-must-gather}
+BASE_COLLECTION_PATH="/must-gather"
```
Comment (dgn) [inline gather_istio:18]: upstream defaults to `$PWD/must-gather`, is this intentional?

File: gather_istio [hunk 4]
```diff
@@ -77,40 +78,89 @@ function getCRDs() {
...
-  local resources="ns/istio-operator MutatingWebhookConfiguration ValidatingWebhookConfiguration"
+		echo "Collecting resources for namespace ${cp}"
+    local pilotName=$(oc get pods -n $namespace -lapp=pilot  -o jsonpath="{.items[0].metadata.name}")
 
-  local controlPlanes="$(getControlPlanes)"
-  resources+="$(addResourcePrefix ns ${controlPlanes})"
+    echo "Overall Envoy synchronization status for namespace ${namespace}" 3>&1 1>${logPath}/controlPlaneStatus
+    local logPath=${BASE_COLLECTION_PATH}/namespaces/${namespace}/controlplane
+    mkdir -p ${logPath}
+    oc exec -it ${pilotName} -n ${namespace} -c discovery -- /usr/local/bin/pilot-discovery request GET /debug/syncz 3>&1 1>${logPath}/synchronization
```
Comment (dgn) [inline gather_istio]: Should we really request an interactive terminal here?

File: gather_istio [hunk 5]
```diff
@@ -77,40 +78,89 @@ function getCRDs() {
...
+    pods="$(oc get pods -n ${podNamespace} -o jsonpath='{ .items[*].metadata.name }')"
+    for podName in ${pods}; do
+        if oc get pod -o yaml ${podName} -n ${podNamespace} | grep -q proxyv2; then
+					  echo "Collecting config for pod ${podName}.${podNamespace}"
+
+            local logPath=${BASE_COLLECTION_PATH}/namespaces/${podNamespace}/pods/${podName}
+            mkdir -p ${logPath}
+
+            echo "Pilot config for pod ${podName}.${podNamespace} from pilot ${pilotName}.${controlPlaneNamespace}" 3>&1 1>${logPath}/pilotConfiguration
+            oc exec -it ${pilotName} -n ${controlPlaneNamespace} -c discovery -- bash -c "/usr/local/bin/pilot-discovery request GET /debug/config_dump?proxyID=${podName}.${podNamespace}" 3>&1 1>${logPath}/pilotConfiguration
```
Comment (dgn) [inline gather_istio]: Again, I don't think we need `-it` here

File: gather_istio [hunk 6]
```diff
@@ -77,40 +78,89 @@ function getCRDs() {
...
+
+            echo "Envoy config for pod ${podName}.${podNamespace} from pilot ${pilotName}.${controlPlaneNamespace}" 3>&1 1>${logPath}/envoyConfiguration
+            oc exec -n ${podNamespace} -it ${podName} -c istio-proxy -- /usr/local/bin/pilot-agent request GET config_dump 3>&1 1>${logPath}/envoyConfiguration
```
Comment (dgn) [inline gather_istio]: same as above

File: gather_istio (hunk 3 above)
Comment (brian-avery) [reply gather_istio:18]: Must-gather actually hardcodes /must-gather in scripts and specifies $PWD/must-gather in others. I reached out to their developers and they said scripts should be outputting to /must-gather as shown in the scripts

**Review by jwendell — CHANGES_REQUESTED** [review]
You are basically changing every line on the script (space x tabs?). Could you please keep whatever the current indentation is so that we can review only the relevant changes?

**Comment by brian-avery** [comment]
Yep. Let me see how to turn off auto-indentation.

**Comment by brian-avery** [comment]
@jwendell removed auto-formatting as per request
@dgn removing -it still works so removed that as well.

File: gather_istio (hunk 3 above)
Comment (brian-avery) [reply gather_istio:18]: https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs#L2
https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_audit_logs#L2
https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs#L2

File: gather_istio [hunk 7]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
   echo ${result}
 }
 
+function getSynchronization() {
+  local namespace="${1}"
+
+  echo "Collecting resources for namespace ${cp}"
+  local pilotName=$(oc get pods -n $namespace -lapp=pilot  -o jsonpath="{.items[0].metadata.name}")
```
Comment (jwendell) [inline gather_istio]: can you surround variables with ${}?

File: gather_istio [hunk 8]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
...
+
+  echo "Overall Envoy synchronization status for namespace ${namespace}" 3>&1 1>${logPath}/controlPlaneStatus
+  local logPath=${BASE_COLLECTION_PATH}/namespaces/${namespace}/controlplane
```
Comment (jwendell) [inline gather_istio:94]: I see you're using the pattern `command 3>&1 1> file` a lot. Any reason for that? Why not just `command > file`?

File: gather_istio [hunk 9]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
   echo ${result}
 }
 
+function getSynchronization() {
```
Comment (jwendell) [inline gather_istio:87]: Can you write a doc/comment on that function briefly explaining what it does and what are the required parameters?

File: gather_istio [hunk 10]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
...
+  echo "Collecting resources for namespace ${cp}"
+  local pilotName=$(oc get pods -n $namespace -lapp=pilot  -o jsonpath="{.items[0].metadata.name}")
+
+  echo "Overall Envoy synchronization status for namespace ${namespace}" 3>&1 1>${logPath}/controlPlaneStatus
+  local logPath=${BASE_COLLECTION_PATH}/namespaces/${namespace}/controlplane
+  mkdir -p ${logPath}
+  oc exec ${pilotName} -n ${namespace} -c discovery -- /usr/local/bin/pilot-discovery request GET /debug/syncz 3>&1 1>${logPath}/synchronization
+}
+
+function getEnvoyConfigForPodsInNamespace() {
```
Comment (jwendell) [inline gather_istio:106]: Can you write a doc/comment on that function briefly explaining what it does and what are the required parameters?

File: gather_istio [hunk 11]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
...
+  local controlPlaneNamespace="${1}"
+  local pilotName=$(oc get pods -n $controlPlaneNamespace -lapp=pilot  -o jsonpath="{.items[0].metadata.name}")
+  local podNamespace="${2}"
+
+  echo "Collecting Envoy config for pods in ${podNamespace} control plane ${controlPlaneNamespace}"
+
```
Comment (jwendell) [inline gather_istio:112]: extra line

File: gather_istio [hunk 12]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
...
+
+  pods="$(oc get pods -n ${podNamespace} -o jsonpath='{ .items[*].metadata.name }')"
```
Comment (jwendell) [inline gather_istio]: local variable?

File: gather_istio [hunk 13]
```diff
@@ -82,7 +120,16 @@ function main() {
   resources+="$(addResourcePrefix ns ${controlPlanes})"
 
   for cp in ${controlPlanes}; do
-    resources+="$(addResourcePrefix ns $(getMembers ${cp}))"
+    local members=$(getMembers ${cp})
+      resources+="$(addResourcePrefix ns $members)"
```
Comment (jwendell) [inline gather_istio]: ${}

File: gather_istio [hunk 14]
```diff
@@ -82,7 +120,16 @@ function main() {
...
+      getSynchronization ${cp}
+
+       #collect Envoy configuration first from control plane pods and then from members
+       getEnvoyConfigForPodsInNamespace ${cp} ${cp}
+       for member in $members; do
```
Comment (jwendell) [inline gather_istio]: ${}

File: gather_istio [hunk 15]
```diff
@@ -71,6 +72,43 @@ function getCRDs() {
...
+  mkdir -p ${logPath}
+  oc exec ${pilotName} -n ${namespace} -c discovery -- /usr/local/bin/pilot-discovery request GET /debug/syncz 3>&1 1>${logPath}/synchronization
+}
+
+function getEnvoyConfigForPodsInNamespace() {
+  local controlPlaneNamespace="${1}"
+  local pilotName=$(oc get pods -n $controlPlaneNamespace -lapp=pilot  -o jsonpath="{.items[0].metadata.name}")
+  local podNamespace="${2}"
+
+  echo "Collecting Envoy config for pods in ${podNamespace} control plane ${controlPlaneNamespace}"
```
Comment (jwendell) [inline gather_istio]: .. for pods in namespace ${podNamespace}, control plane...

**Review by jwendell — CHANGES_REQUESTED** [review]

File: gather_istio (hunk 8 above)
Comment (brian-avery) [reply gather_istio:94]: I want to catch any errors if they occur as well.

Comment (brian-avery) [reply gather_istio:94]: 0 - stdin
1 - stdout
2 - stderr

That should be 2>&1 1>file

**Review by dgn — APPROVED** [review]

**Review by knrc — APPROVED** [review]

**Review by jwendell — APPROVED** [review]
LGTM, thanks.

**Pull request merged**